tf_quant_finance/black_scholes/asian_prices.py

```diff
@@ -0,0 +1,218 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Black Scholes prices of a batch of Asian options."""
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+from tf_quant_finance.black_scholes import vanilla_prices
+from tf_quant_finance import types
+__all__ = [
+    'discrete_geometric_average_price'
+]
+
+
+def discrete_geometric_average_price(*,
+                                     volatilities: types.RealTensor,
+                                     strikes: types.RealTensor,
+                                     expiries: types.RealTensor,
+                                     spots: types.RealTensor = None,
+                                     forwards: types.RealTensor = None,
+                                     sampling_times: types.RealTensor = None,
+                                     past_fixings: types.RealTensor = None,
+                                     discount_rates: types.RealTensor = None,
+                                     dividend_rates: types.RealTensor = None,
+                                     discount_factors: types.RealTensor = None,
+                                     is_call_options: types.BoolTensor = None,
+                                     is_normal_volatility: bool = False,
+                                     dtype: tf.DType = None,
+                                     name: str = None) -> types.RealTensor:
+  """Computes the Black Scholes price for a batch of geometric averaging,
+  discrete time sampled call or put options.
+
+  In Black-Scholes, the marginal distribution of the underlying at each sampling
+  date is lognormal. The product of a sequence of lognormal variables is also
+  lognormal so we can re-express these options as vanilla options with modified
+  parameters and use the vanilla pricer to price them.
+
+  TODO: support volatility termstructures
+  For now we provide a single volatility for each option. A simple extension of
+  Black-Scholes provides a termstructure of volatilities that obeys non-negative
+  forward variance. We could also use this methodology under that model.
+
+  #### Example
+
+  ```python
+    # Price a batch of 5 seasoned discrete geometric asian options.
+    volatilities = np.array([0.0001, 102.0, 2.0, 0.1, 0.4])
+    forwards = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    # Strikes will automatically be broadcasted to shape [5].
+    strikes = np.array([3.0])
+    # Expiries will be broadcast to shape [5], i.e. each option has strike=3
+    # and expiry = 1.
+    expiries = 1.0
+    sampling_times = np.array([[0.5, 0.5, 0.5, 0.5, 0.5], [1.0, 1.0, 1.0, 1.0, 1.0]])
+    past_fixings = np.array([[1.0, 2.0, 3.0, 4.0, 5.0]])
+    computed_prices = tff.black_scholes.discrete_geometric_average_price(
+        volatilities=volatilities,
```

Comment:
I think `discrete_geometric_average_price` is a bit vague without referencing Asian options. How about simply  `asian_option_price`? You could add enums for aggregation type as well as option type (discrete/continuous)

Reply:
Great feedback - this is done now. Pricing for more variants coming soon.